diff --git a/cmake/Android/GradleClean.cmake b/cmake/Android/GradleClean.cmake
index f8659420..d2ad32af 100644
--- a/cmake/Android/GradleClean.cmake
+++ b/cmake/Android/GradleClean.cmake
@@ -23,7 +23,14 @@
 configure_file("${LINPHONESDK_DIR}/cmake/Android/gradle/build.gradle.cmake" "${LINPHONESDK_BUILD_DIR}/build.gradle" @ONLY)
 configure_file("${LINPHONESDK_DIR}/cmake/Android/gradle/upload.gradle.cmake" "${LINPHONESDK_BUILD_DIR}/upload.gradle" @ONLY)
 configure_file("${LINPHONESDK_DIR}/cmake/Android/gradle/gradle.properties.cmake" "${LINPHONESDK_BUILD_DIR}/gradle.properties" @ONLY)
-configure_file("${LINPHONESDK_DIR}/cmake/Android/gradle/LinphoneSdkManifest.xml.cmake" "${LINPHONESDK_BUILD_DIR}/LinphoneSdkManifest.xml" @ONLY)
+
+if(ENABLE_VIDEO)
+	message(STATUS "Build with video, keeping CAMERA and FOREGROUND_SERVICE_CAMERA permissions in Android Manifest")
+	configure_file("${LINPHONESDK_DIR}/cmake/Android/gradle/LinphoneSdkManifest.xml.cmake" "LinphoneSdkManifest.xml" @ONLY)
+else()
+	message(STATUS "Build without video, removing CAMERA and FOREGROUND_SERVICE_CAMERA permissions from Android Manifest")
+	configure_file("${LINPHONESDK_DIR}/cmake/Android/gradle/LinphoneSdkManifestNoVideo.xml.cmake" "LinphoneSdkManifest.xml" @ONLY)
+endif()
 
 execute_process(
 	COMMAND "${LINPHONESDK_DIR}/cmake/Android/gradlew" "clean"
diff --git a/cmake/LinphoneSdkPlatformAndroid.cmake b/cmake/LinphoneSdkPlatformAndroid.cmake
index 20ab876a..6103fa54 100644
--- a/cmake/LinphoneSdkPlatformAndroid.cmake
+++ b/cmake/LinphoneSdkPlatformAndroid.cmake
@@ -31,7 +31,7 @@ list(GET _archs 0 _first_arch)
 
 
 add_custom_target(gradle-clean ALL
-	"${CMAKE_COMMAND}" "-DLINPHONESDK_DIR=${LINPHONESDK_DIR}" "-DLINPHONESDK_BUILD_DIR=${CMAKE_BINARY_DIR}" "-DLINPHONESDK_VERSION=${LINPHONESDK_VERSION}" "-DLINPHONESDK_STATE=${LINPHONESDK_STATE}" "-DLINPHONESDK_BRANCH=${LINPHONESDK_BRANCH}" "-DLINPHONESDK_FIRST_ARCH=${_first_arch}" "-P" "${LINPHONESDK_DIR}/cmake/Android/GradleClean.cmake"
+	"${CMAKE_COMMAND}" "-DLINPHONESDK_DIR=${LINPHONESDK_DIR}" "-DLINPHONESDK_BUILD_DIR=${CMAKE_BINARY_DIR}" "-DLINPHONESDK_VERSION=${LINPHONESDK_VERSION}" "-DLINPHONESDK_STATE=${LINPHONESDK_STATE}" "-DLINPHONESDK_BRANCH=${LINPHONESDK_BRANCH}" "-DLINPHONESDK_FIRST_ARCH=${_first_arch}" "-DENABLE_VIDEO=${ENABLE_VIDEO}" "-P" "${LINPHONESDK_DIR}/cmake/Android/GradleClean.cmake"
 )
 list(APPEND LINPHONESDK_PREBUILD_DEPENDENCIES gradle-clean)
 
diff --git a/mediastreamer2/java/src/org/linphone/mediastream/Version.java b/mediastreamer2/java/src/org/linphone/mediastream/Version.java
index 67030169..8473077b 100644
--- a/mediastreamer2/java/src/org/linphone/mediastream/Version.java
+++ b/mediastreamer2/java/src/org/linphone/mediastream/Version.java
@@ -65,6 +65,7 @@ public class Version {
 	public static final int API32_ANDROID_12L = 32;
 	public static final int API33_ANDROID_13_TIRAMISU = 33;
 	public static final int API34_ANDROID_14_UPSIDE_DOWN_CAKE = 34;
+	public static final int API35_ANDROID_15_VANILLA_ICE_CREAM = 35;
 
 	private static native boolean nativeHasZrtp();
 	private static native boolean nativeHasNeon();
diff --git a/cmake/Android/gradle/LinphoneSdkManifestNoVideo.xml.cmake b/cmake/Android/gradle/LinphoneSdkManifestNoVideo.xml.cmake
new file mode 100644
index 00000000..deca08bc
--- /dev/null
+++ b/cmake/Android/gradle/LinphoneSdkManifestNoVideo.xml.cmake
@@ -0,0 +1,30 @@
+<?xml version="1.0" encoding="utf-8"?>
+<manifest package="org.linphone.core"
+    xmlns:android="http://schemas.android.com/apk/res/android"
+    android:installLocation="auto">
+
+    <!-- SDK required permissions -->
+    <uses-permission android:name="android.permission.INTERNET" />
+    <uses-permission android:name="android.permission.RECORD_AUDIO" />
+    <uses-permission android:name="android.permission.ACCESS_NETWORK_STATE" />
+    <uses-permission android:name="android.permission.WAKE_LOCK" />
+    <!-- Needed to be able to use WifiManager.MulticastLock -->
+    <uses-permission android:name="android.permission.CHANGE_WIFI_MULTICAST_STATE" />
+    <!-- Needed for bluetooth -->
+    <uses-permission android:name="android.permission.BLUETOOTH" />
+    <!-- Needed for bluetooth headset -->
+    <uses-permission android:name="android.permission.MODIFY_AUDIO_SETTINGS" />
+    <!-- To vibrate while incoming call -->
+    <uses-permission android:name="android.permission.VIBRATE" />
+    <!-- Needed to check current do not disturb policy -->
+    <uses-permission android:name="android.permission.ACCESS_NOTIFICATION_POLICY" />
+    <!-- Needed to gain access to network temporarily when in background on a metered network if data saver is on when a push is being received -->
+    <!-- Needed for Android 14
+    https://developer.android.com/about/versions/14/behavior-changes-14#fgs-types -->
+    <uses-permission android:name="android.permission.FOREGROUND_SERVICE"/>
+    <uses-permission android:name="android.permission.FOREGROUND_SERVICE_MICROPHONE" />
+    <uses-permission android:name="android.permission.FOREGROUND_SERVICE_PHONE_CALL" />
+    <uses-permission android:name="android.permission.FOREGROUND_SERVICE_DATA_SYNC" />
+    <!-- End of SDK required permissions -->
+	
+</manifest>
