diff --git a/liblinphone/src/nat/ice-service.cpp b/liblinphone/src/nat/ice-service.cpp
index a559808b9..3e5f20297 100644
--- a/liblinphone/src/nat/ice-service.cpp
+++ b/liblinphone/src/nat/ice-service.cpp
@@ -259,7 +259,8 @@ int IceService::gatherLocalCandidates() {
 		localAddrs.push_back(mediaLocalIp);
 	}
 
-#if defined(__APPLE__) && TARGET_OS_IPHONE
+// TN patch - disable local network permission check, looks like we don't need it for TCP TURN
+#if 0
 	if (getPlatformHelpers(getCCore())->getNetworkType() == PlatformHelpers::NetworkType::Wifi &&
 	    !hasLocalNetworkPermission(localAddrs))
 		return -1;
