diff --git a/liblinphone/coreapi/linphonecore.c b/liblinphone/coreapi/linphonecore.c
index f5b148842..4b50cd47b 100644
--- a/liblinphone/coreapi/linphonecore.c
+++ b/liblinphone/coreapi/linphonecore.c
@@ -8269,6 +8269,10 @@ void linphone_core_activate_audio_session(LinphoneCore *lc, bool_t actived) {
 	L_GET_CPP_PTR_FROM_C_OBJECT(lc)->soundcardActivateAudioSession(actived);
 }
 
+bool_t linphone_core_is_audio_session_active(LinphoneCore* lc) {
+	return L_GET_CPP_PTR_FROM_C_OBJECT(lc)->soundcardIsAudioSessionActive();
+}
+
 void linphone_core_configure_audio_session(LinphoneCore *lc) {
 	if (linphone_core_get_calls_nb(lc) < 2) {
 		L_GET_CPP_PTR_FROM_C_OBJECT(lc)->soundcardConfigureAudioSession();
diff --git a/liblinphone/include/linphone/core_utils.h b/liblinphone/include/linphone/core_utils.h
index a4d572cf0..571b51574 100644
--- a/liblinphone/include/linphone/core_utils.h
+++ b/liblinphone/include/linphone/core_utils.h
@@ -133,6 +133,14 @@ LINPHONE_PUBLIC bool_t linphone_core_is_echo_canceller_calibration_required(Linp
  */
 LINPHONE_PUBLIC void linphone_core_start_dtmf_stream(LinphoneCore *core);
 
+/**
+ * Special function to indicate if the audio session is activated.
+ * @param core The #LinphoneCore object. @notnil
+ * @return TRUE to if activated, FALSE otherwise.
+ * @ingroup IOS
+ */
+LINPHONE_PUBLIC bool_t linphone_core_is_audio_session_active(LinphoneCore* core);
+
 /**
  * @ingroup IOS
  * @param core The #LinphoneCore object. @notnil
@@ -149,6 +157,14 @@ LINPHONE_PUBLIC void linphone_core_stop_dtmf_stream(LinphoneCore *core);
  */
 LINPHONE_PUBLIC void linphone_core_activate_audio_session(LinphoneCore *core, bool_t actived);
 
+/**
+ * Indicates if the audio session is activated.
+ * @param core The #LinphoneCore object. @notnil
+ * @return TRUE to if activated, FALSE otherwise.
+ * @ingroup IOS
+ */
+LINPHONE_PUBLIC bool_t linphone_core_is_audio_session_active(LinphoneCore* core);
+
 /**
  * Special function to configure audio session with default settings. Must be called in ProviderDelegate's callbacks
  * when answer an incoming call and start an outgoing call.
diff --git a/liblinphone/src/core/core-call.cpp b/liblinphone/src/core/core-call.cpp
index c7f5e4de6..24cc82cef 100644
--- a/liblinphone/src/core/core-call.cpp
+++ b/liblinphone/src/core/core-call.cpp
@@ -311,6 +311,14 @@ void Core::soundcardActivateAudioSession(bool actived) {
 	}
 }
 
+bool Core::soundcardIsAudioSessionActive() {
+	MSSndCard *card = getCCore()->sound_conf.play_sndcard;
+	if (card) {
+		return ms_snd_card_is_audio_session_active(card);
+	}
+	return false;
+}
+
 void Core::soundcardConfigureAudioSession() {
 	MSSndCard *card = getCCore()->sound_conf.play_sndcard;
 	if (card) {
diff --git a/liblinphone/src/core/core.h b/liblinphone/src/core/core.h
index 1b3a4634e..2d8c83768 100644
--- a/liblinphone/src/core/core.h
+++ b/liblinphone/src/core/core.h
@@ -162,6 +162,7 @@ public:
 	std::shared_ptr<Call> getCurrentCall() const;
 	LinphoneStatus pauseAllCalls();
 	void soundcardActivateAudioSession(bool active);
+	bool soundcardIsAudioSessionActive();
 	void soundcardConfigureAudioSession();
 	void soundcardEnableCallkit(bool enabled);
 	void soundcardAudioRouteChanged();
diff --git a/mediastreamer2/include/mediastreamer2/mssndcard.h b/mediastreamer2/include/mediastreamer2/mssndcard.h
index 0293cd4e..f3fc6934 100644
--- a/mediastreamer2/include/mediastreamer2/mssndcard.h
+++ b/mediastreamer2/include/mediastreamer2/mssndcard.h
@@ -91,6 +91,7 @@ typedef struct _MSSndCard *(*MSSndCardDuplicateFunc)(struct _MSSndCard *obj);
 typedef void (*MSSndCardSetUsageHintFunc)(struct _MSSndCard *obj, bool_t is_going_to_be_used);
 typedef void (*MSSndCardUnloadFunc)(MSSndCardManager *obj);
 typedef void (*MSSndCardAudioSessionFunc)(struct _MSSndCard *obj, bool_t actived);
+typedef bool_t (*MSSndCardIsAudioSessionActiveFunc)(struct _MSSndCard *obj);
 typedef void (*MSSndCardCallKitFunc)(struct _MSSndCard *obj, bool_t enabled);
 typedef void (*MSSndCardAudioRouteFunc)(struct _MSSndCard *obj);
 typedef void (*MSSndCardConfigureFunc)(struct _MSSndCard *obj);
@@ -111,6 +112,7 @@ struct _MSSndCardDesc {
 	MSSndCardUnloadFunc unload;
 	MSSndCardSetUsageHintFunc usage_hint;
 	MSSndCardAudioSessionFunc audio_session_activated;
+	MSSndCardIsAudioSessionActiveFunc audio_session_is_active;
 	MSSndCardCallKitFunc callkit_enabled;
 	MSSndCardAudioRouteFunc audio_route_changed;
 	MSSndCardConfigureFunc configure;
@@ -741,6 +743,14 @@ MS2_PUBLIC void ms_snd_card_set_usage_hint(MSSndCard *obj, bool_t is_going_to_be
  */
 MS2_PUBLIC void ms_snd_card_notify_audio_session_activated(MSSndCard *obj, bool_t activated);
 
+/**
+ * Indicates if the audio session is activated. See ms_snd_card_notify_audio_session_activated()
+ *
+ * @param obj      A sound card object.
+ * @return         TRUE if audio session is activated, FALSE otherwise.
+ */
+MS2_PUBLIC bool_t ms_snd_card_is_audio_session_active(MSSndCard *obj);
+
 /**
  * Used by application to notify whether audio route is changed. On most platform this function is useless.
  * But  an iOS application needs to explicitely notify mediastreamer2 with ms_snd_card_notify_audio_route_changed()
diff --git a/mediastreamer2/src/audiofilters/msiounit.mm b/mediastreamer2/src/audiofilters/msiounit.mm
index f60d3075..ea0e99f2 100644
--- a/mediastreamer2/src/audiofilters/msiounit.mm
+++ b/mediastreamer2/src/audiofilters/msiounit.mm
@@ -940,6 +940,11 @@ static void au_audio_route_changed(MSSndCard *obj) {
 	handle_sample_rate_change(FALSE);
 }
 
+static bool_t au_audio_session_is_active(MSSndCard *obj) {
+	AudioUnitHolder *au_holder = [AudioUnitHolder sharedInstance];
+	return au_holder.audio_session_activated;
+}
+
 static void au_audio_session_activated(MSSndCard *obj, bool_t activated) {
 	AudioUnitHolder *au_holder = [AudioUnitHolder sharedInstance];
 	bool_t need_audio_session_reconfiguration = FALSE;
@@ -1030,6 +1035,7 @@ static void au_configure(MSSndCard *obj) {
 .duplicate=au_duplicate,
 .usage_hint=au_usage_hint,
 .audio_session_activated=au_audio_session_activated,
+.audio_session_is_active=au_audio_session_is_active,
 .callkit_enabled=au_callkit_enabled,
 .audio_route_changed=au_audio_route_changed,
 .configure=au_configure,
diff --git a/mediastreamer2/src/base/mssndcard.c b/mediastreamer2/src/base/mssndcard.c
index f6000b78..6a16a774 100644
--- a/mediastreamer2/src/base/mssndcard.c
+++ b/mediastreamer2/src/base/mssndcard.c
@@ -509,6 +509,12 @@ void ms_snd_card_notify_audio_session_activated(MSSndCard *obj, bool_t activated
 	if (obj->desc->audio_session_activated != NULL) obj->desc->audio_session_activated(obj, activated);
 }
 
+bool_t ms_snd_card_is_audio_session_active(MSSndCard *obj) {
+	if (obj->desc->audio_session_is_active != NULL)
+		return obj->desc->audio_session_is_active(obj);
+	return FALSE;
+}
+
 void ms_snd_card_notify_audio_route_changed(MSSndCard *obj) {
 	if (obj->desc->audio_route_changed != NULL) obj->desc->audio_route_changed(obj);
 }
