diff --git a/liblinphone/share/CMakeLists.txt b/liblinphone/share/CMakeLists.txt
index 60d4c00b7..741691090 100644
--- a/liblinphone/share/CMakeLists.txt
+++ b/liblinphone/share/CMakeLists.txt
@@ -36,9 +36,9 @@ if(ENABLE_ASSETS)
 		"${CMAKE_CURRENT_SOURCE_DIR}/hello8000.wav"
 		"${CMAKE_CURRENT_SOURCE_DIR}/incoming_chat.wav"
 		"${CMAKE_CURRENT_SOURCE_DIR}/ringback.wav"
-		"${CMAKE_CURRENT_SOURCE_DIR}/dont_wait_too_long.mkv"
+		# "${CMAKE_CURRENT_SOURCE_DIR}/dont_wait_too_long.mkv"
 		"${CMAKE_CURRENT_SOURCE_DIR}/toy-mono.wav"
-		"${CMAKE_CURRENT_SOURCE_DIR}/silence.mkv"
+		# "${CMAKE_CURRENT_SOURCE_DIR}/silence.mkv"
 		PARENT_SCOPE
 	)
 else()
@@ -61,7 +61,7 @@ if(ENABLE_GTK_UI)
 	)
 endif()
 
-if(ENABLE_ASSETS)
-	add_subdirectory(rings)
-	set(RING_FILES ${RING_FILES} PARENT_SCOPE)
-endif()
+# if(ENABLE_ASSETS)
+# 	add_subdirectory(rings)
+# 	set(RING_FILES ${RING_FILES} PARENT_SCOPE)
+# endif()
