diff --git a/liblinphone/src/conference/client-conference.cpp b/liblinphone/src/conference/client-conference.cpp
index f9c2143af..bb3a6b314 100644
--- a/liblinphone/src/conference/client-conference.cpp
+++ b/liblinphone/src/conference/client-conference.cpp
@@ -936,12 +936,13 @@ void ClientConference::onFocusCallStateChanged(CallSession::State state, BCTBX_U
 					if (requestStreams() != 0) {
 						lInfo() << "Delaying re-INVITE in order to get streams after joining " << *this
 						        << " because the dialog is not available yet to accept this transaction";
-						session->addPendingAction(requestStreams);
+						mScheduleUpdate = true;
+					} else {
+						// An update has been successfully sent therefore clear the flag mScheduleUpdate to avoid
+						// sending it twice.
+						mScheduleUpdate = false;
+						mFullStateUpdate = false;
 					}
-					// An update has been successfully sent therefore clear the flag mScheduleUpdate to avoid sending it
-					// twice.
-					mScheduleUpdate = false;
-					mFullStateUpdate = false;
 				}
 			}
 				BCTBX_NO_BREAK; /* Intentional no break */
@@ -1793,6 +1794,13 @@ void ClientConference::onFullStateReceived() {
 		}
 #endif // HAVE_ADVANCED_IM
 
+		auto requestStreams = [this]() -> LinphoneStatus {
+			lInfo() << "Sending re-INVITE in order to get streams after receiving a NOTIFY full state for " << *this;
+			setState(ConferenceInterface::State::Created);
+			auto ret = updateMainSession(false);
+			return ret;
+		};
+
 		auto session = mFocus ? dynamic_pointer_cast<MediaSession>(mFocus->getSession()) : nullptr;
 		// Notify local participant that the microphone is muted when receiving the full state as participants are added
 		// as soon as possible
@@ -1806,22 +1814,12 @@ void ClientConference::onFullStateReceived() {
 
 		if (!getCore()->getCCore()->sal->mediaDisabled()) {
 			if (session && (!session->mediaInProgress() || !session->getPrivate()->isUpdateSentWhenIceCompleted())) {
-
-				auto requestStreams = [this]() -> LinphoneStatus {
-					lInfo() << "Sending re-INVITE in order to get streams after receiving a NOTIFY full state for "
-					        << *this;
-					setState(ConferenceInterface::State::Created);
-					auto ret = updateMainSession(false);
-					return ret;
-				};
-
 				if (requestStreams() != 0) {
 					lInfo() << "Delaying re-INVITE in order to get streams after receiving a NOTIFY full state for "
 					        << *this << " because it cannot be sent right now";
-					session->addPendingAction(requestStreams);
+					mScheduleUpdate = true;
+					mFullStateUpdate = true;
 				}
-				mScheduleUpdate = false;
-				mFullStateUpdate = false;
 			} else {
 				lInfo() << "Delaying re-INVITE in order to get streams after receiving a NOTIFY full state for "
 				        << *this << " because ICE negotiations didn't end yet";
diff --git a/liblinphone/src/conference/session/call-session.cpp b/liblinphone/src/conference/session/call-session.cpp
index cfc914c78..cec0981a7 100644
--- a/liblinphone/src/conference/session/call-session.cpp
+++ b/liblinphone/src/conference/session/call-session.cpp
@@ -803,14 +803,6 @@ bool CallSessionPrivate::isReadyForInvite() const {
 
 bool CallSessionPrivate::isUpdateAllowed(CallSession::State &nextState) const {
 	L_Q();
-	if (op->hasRetryFunction()) {
-		lWarning() << "Unable to send reINVITE or UPDATE request right now because " << *q << " (local address "
-		           << *q->getLocalAddress() << " remote address "
-		           << (q->getRemoteAddress() ? q->getRemoteAddress()->toString() : "sip:")
-		           << ") needs to execute the request which was replied by a 491 Request Pending first.";
-		return false;
-	}
-
 	switch (state) {
 		case CallSession::State::IncomingReceived:
 		case CallSession::State::PushIncomingReceived:
@@ -1192,40 +1184,15 @@ CallSessionPrivate::ContactInfo CallSessionPrivate::chooseContact() const {
 
 void CallSessionPrivate::reinviteToRecoverFromConnectionLoss() {
 	L_Q();
-	lInfo() << *q << " is going to be updated (reINVITE) in order to recover from lost connectivity";
-	if (op) {
-		// Reset retry function as we need to recover from a network loss
-		op->resetRetryFunction();
-	}
+	lInfo() << "CallSession [" << q << "] is going to be updated (reINVITE) in order to recover from lost connectivity";
 	q->update(mParams, CallSession::UpdateMethod::Invite);
 }
 
 void CallSessionPrivate::repairByNewInvite(bool withReplaces) {
 	L_Q();
-
-	// Default Replaces header (RFC3891) to the sal setting. Nonetheless if the supported header of the account should
-	// be used, then override it
-	bool replacesSupported = op->getSal()->hasSupportedTag("replaces");
-	const auto &account = getDestAccount();
-	if (account) {
-		const auto &accountParams = account->getAccountParams();
-		if (accountParams && accountParams->useSupportedTags()) {
-			replacesSupported = accountParams->hasSupportedTag("replaces");
-		}
-	}
-
-	if (withReplaces && !replacesSupported) {
-		lInfo() << "Terminate " << *q
-		        << " because it is not possible to recover it if the Replaces header (RFC3891) is not supported";
-		terminate();
-		return;
-	}
-
-	lInfo() << *q << " is going to have a new INVITE one in order to recover from lost connectivity; "
-	        << (withReplaces ? "with" : "without") << " Replaces header";
-
-	// Reset retry function as we need to repair the INVITE session
-	op->resetRetryFunction();
+	lInfo() << "CallSession [" << q
+	        << "] is going to have a new INVITE one in order to recover from lost connectivity; with Replaces header:"
+	        << (withReplaces ? "yes" : "no");
 
 	// FIXME: startInvite shall() accept a list of bodies.
 	// Since it is not the case, we can only re-use the first one.
diff --git a/liblinphone/src/conference/session/media-session.cpp b/liblinphone/src/conference/session/media-session.cpp
index a985d7f1e..9a4535497 100644
--- a/liblinphone/src/conference/session/media-session.cpp
+++ b/liblinphone/src/conference/session/media-session.cpp
@@ -4418,10 +4418,6 @@ void MediaSessionPrivate::reinviteToRecoverFromConnectionLoss() {
 	lInfo() << "MediaSession [" << q
 	        << "] is going to be updated (reINVITE) in order to recover from lost connectivity";
 	getStreamsGroup().getIceService().resetSession();
-	if (op) {
-		// Reset retry function as we need to recover from a network loss
-		op->resetRetryFunction();
-	}
 	MediaSessionParams newParams(*getParams());
 	q->update(&newParams, CallSession::UpdateMethod::Invite, q->isCapabilityNegotiationEnabled());
 }
diff --git a/liblinphone/src/sal/call-op.cpp b/liblinphone/src/sal/call-op.cpp
index a5519f962..841bca102 100644
--- a/liblinphone/src/sal/call-op.cpp
+++ b/liblinphone/src/sal/call-op.cpp
@@ -528,9 +528,6 @@ void SalCallOp::processResponseCb(void *userCtx, const belle_sip_response_event_
 	lInfo() << "Op [" << op << "] receiving call response [" << code << "], dialog is [" << dialog << "] in state ["
 	        << belle_sip_dialog_state_to_string(dialogState) << "]";
 	op->ref(); // To make sure no cb will destroy op
-	if (code != 491) {
-		op->resetRetryFunction(); // Retry function has been either executed or not needed anymore
-	}
 
 	auto request = belle_sip_transaction_get_request(BELLE_SIP_TRANSACTION(clientTransaction));
 	string method = belle_sip_request_get_method(request);
