diff --git a/src/mediastreamer2/audiofilters/msiounit.mm b/mediastreamer2/src/audiofilters/msiounit.mm
index 9c0ded73..0ad4e2d4 100644
--- a/mediastreamer2/src/audiofilters/msiounit.mm
+++ b/mediastreamer2/src/audiofilters/msiounit.mm
@@ -924,7 +924,7 @@ static void au_audio_session_activated(MSSndCard *obj, bool_t activated) {
 			*/
 			handle_sample_rate_change(need_audio_session_reconfiguration);
 			/* The next is done on a separate thread because it is considerably slow, so don't block the application calling thread here. */
-			dispatch_async(dispatch_get_global_queue(DISPATCH_QUEUE_PRIORITY_HIGH,0), ^{
+			dispatch_async(dispatch_get_main_queue(), ^{
 					[[AudioUnitHolder sharedInstance] configure_audio_unit];
 					[[AudioUnitHolder sharedInstance] start_audio_unit:0];
 			});
