diff --git a/liblinphone/coreapi/linphonecore.c b/liblinphone/coreapi/linphonecore.c
index 898c26d78..9a0153d2d 100644
--- a/liblinphone/coreapi/linphonecore.c
+++ b/liblinphone/coreapi/linphonecore.c
@@ -6181,6 +6181,18 @@ MSFactory *linphone_core_get_ms_factory(LinphoneCore *lc) {
 	return lc->factory;
 }
 
+// TN patch
+extern unsigned char linphone_tn_terminate_on_cancel_enabled;
+
+bool_t linphone_core_get_terminate_on_cancel_enabled(const LinphoneCore */*core*/) {
+	return linphone_tn_terminate_on_cancel_enabled;
+}
+
+void linphone_core_set_terminate_on_cancel_enabled(LinphoneCore */*core*/, bool_t enable) {
+	linphone_tn_terminate_on_cancel_enabled = enable;
+}
+// TN patch
+
 void linphone_core_set_ringback(LinphoneCore *lc, const char *path) {
 	if (lc->sound_conf.remote_ring != 0) {
 		ms_free(lc->sound_conf.remote_ring);
diff --git a/liblinphone/include/linphone/core.h b/liblinphone/include/linphone/core.h
index 7ed1fa480..e2371432e 100644
--- a/liblinphone/include/linphone/core.h
+++ b/liblinphone/include/linphone/core.h
@@ -3468,6 +3468,25 @@ LINPHONE_PUBLIC void linphone_core_set_root_ca_data(LinphoneCore *core, const ch
  */
 LINPHONE_PUBLIC void linphone_core_set_ssl_config(LinphoneCore *core, void *ssl_config);
 
+// TN patch
+
+/**
+ * Tells whether we should terminate an established call on a SIP cancel.
+ * @param core #LinphoneCore object @notnil
+ * @ingroup media_parameters
+ */
+bool_t linphone_core_get_terminate_on_cancel_enabled(const LinphoneCore *core);
+
+/**
+ * Tells whether we should terminate an established call on a SIP cancel.
+ * @param core #LinphoneCore object @notnil
+ * @param enable Boolean value telling whether the feature is enabled.
+ * @ingroup media_parameters
+ */
+void linphone_core_set_terminate_on_cancel_enabled(LinphoneCore *core, bool_t enable);
+
+// TN patch
+
 /**
  * Sets the path to a wav file used for ringing back.
  * Ringback means the ring that is heard when it's ringing at the remote party.
diff --git a/liblinphone/src/sal/call-op.cpp b/liblinphone/src/sal/call-op.cpp
index d159497c9..dc599e624 100644
--- a/liblinphone/src/sal/call-op.cpp
+++ b/liblinphone/src/sal/call-op.cpp
@@ -29,6 +29,8 @@
 #include <belle-sip/defs.h>
 #include <belle-sip/provider.h>
 
+unsigned char linphone_tn_terminate_on_cancel_enabled = 0;
+
 using namespace std;
 
 LINPHONE_BEGIN_NAMESPACE
@@ -923,12 +925,16 @@ void SalCallOp::processRequestEventCb(void *userCtx, const belle_sip_request_eve
 		case BELLE_SIP_DIALOG_EARLY:
 			if (method == "CANCEL") {
 				if (belle_sip_request_event_get_server_transaction(event)) {
+					lInfo() << "SIP CANCEL case 1 - pre dialog terminating call";
+
 					// First answer 200 ok to cancel
 					belle_sip_server_transaction_send_response(serverTransaction,
 					                                           op->createResponseFromRequest(request, 200));
 					// Terminate invite transaction
 					op->callTerminated(op->mPendingServerTransaction, 487, request);
 				} else {
+					lInfo() << "SIP CANCEL case 2 - pre dialog call does not exist";
+
 					// Call leg does not exist
 					belle_sip_server_transaction_send_response(serverTransaction,
 					                                           op->createResponseFromRequest(request, 481));
@@ -1028,10 +1034,19 @@ void SalCallOp::processRequestEventCb(void *userCtx, const belle_sip_request_eve
 			} else if (method == "CANCEL") {
 				auto lastTransaction = belle_sip_dialog_get_last_transaction(op->mDialog);
 				if (!lastTransaction || !isAPendingIncomingInviteTransaction(lastTransaction)) {
+					lInfo() << "SIP CANCEL case 3 - post dialog";
+
 					// Call leg does not exist because 200ok already sent
 					belle_sip_server_transaction_send_response(serverTransaction,
 					                                           op->createResponseFromRequest(request, 481));
+
+					if (linphone_tn_terminate_on_cancel_enabled) {
+						lInfo() << "SIP CANCEL case 3 - post dialog terminate on cancel enabled";
+						op->callTerminated(serverTransaction, 200, request);
+					}
 				} else {
+					lInfo() << "SIP CANCEL case 4 - post dialog reINVITE";
+
 					// CANCEL on re-INVITE for which a 200ok has not been sent yet
 					belle_sip_server_transaction_send_response(serverTransaction,
 					                                           op->createResponseFromRequest(request, 200));
