diff --git a/mediastreamer2/src/voip/turn_tcp.cpp b/mediastreamer2/src/voip/turn_tcp.cpp
index 5cac1005..2da7b887 100644
--- a/mediastreamer2/src/voip/turn_tcp.cpp
+++ b/mediastreamer2/src/voip/turn_tcp.cpp
@@ -327,8 +327,8 @@ SslContext::~SslContext() {
 	bctbx_x509_certificate_free(mRootCertificate);
 }
 
-int SslContext::connect() {
-	int error = bctbx_ssl_handshake(mContext);
+int SslContext::connect(bool& isRunning) {
+	int error = bctbx_ssl_handshake2(mContext, (void*)&isRunning);
 	if (error < 0) {
 		char errbuf[1024] = {0};
 		bctbx_strerror(error, errbuf, sizeof(errbuf) - 1);
@@ -419,6 +419,15 @@ int TurnSocket::connect() {
 		return -1;
 	}
 
+	// Set a 2s recv timeout, so the SSL handshake recv can fail gracefully
+	struct timeval recv_tv;
+	recv_tv.tv_sec = 2;
+	recv_tv.tv_usec = 0;
+	error = setsockopt(mSocket, SOL_SOCKET, SO_RCVTIMEO, (char*)&recv_tv, sizeof(recv_tv));
+	if (error != 0) {
+		ms_error("TurnSocket [%p]: setsockopt SO_RCVTIMEO failed: %s", this, getSocketError());
+	}
+
 	// Add HTTP Proxy connection here if needed
 
 	set_blocking_socket(mSocket);
@@ -427,7 +436,7 @@ int TurnSocket::connect() {
 		mSsl =
 		    std::make_unique<SslContext>(mSocket, mClient->mRootCertificatePath, mClient->mTurnServerCn, mClient->mRng);
 
-		error = mSsl->connect();
+		error = mSsl->connect(mRunning);
 		if (error < 0) {
 			ms_error("TurnSocket [%p]: SSL handshake failed", this);
 			mSsl.reset();
@@ -646,7 +655,7 @@ void TurnSocket::runSend() {
 void TurnSocket::runRead() {
 	while (mRunning) {
 		if (mSocket == INVALID_SOCKET) {
-			if (connect() < 0) {
+			if (mRunning && connect() < 0) {
 				ms_usleep(500000);
 			}
 		} else {
diff --git a/mediastreamer2/src/voip/turn_tcp.h b/mediastreamer2/src/voip/turn_tcp.h
index c8ea73ac..3243a371 100644
--- a/mediastreamer2/src/voip/turn_tcp.h
+++ b/mediastreamer2/src/voip/turn_tcp.h
@@ -161,7 +161,7 @@ public:
 	SslContext(const SslContext &) = delete;
 	SslContext(SslContext &&) = delete;
 
-	int connect();
+	int connect(bool& isRunning);
 	int close();
 
 	int read(unsigned char *buffer, size_t length);
diff --git a/bctoolbox/include/bctoolbox/crypto.h b/bctoolbox/include/bctoolbox/crypto.h
index dfebd0d..1c99428 100644
--- a/bctoolbox/include/bctoolbox/crypto.h
+++ b/bctoolbox/include/bctoolbox/crypto.h
@@ -537,6 +537,7 @@ BCTBX_PUBLIC int32_t bctbx_ssl_read(bctbx_ssl_context_t *ssl_ctx, unsigned char
 BCTBX_PUBLIC int32_t bctbx_ssl_write(bctbx_ssl_context_t *ssl_ctx, const unsigned char *buf, size_t buf_length);
 BCTBX_PUBLIC int32_t bctbx_ssl_set_hostname(bctbx_ssl_context_t *ssl_ctx, const char *hostname);
 BCTBX_PUBLIC int32_t bctbx_ssl_handshake(bctbx_ssl_context_t *ssl_ctx);
+BCTBX_PUBLIC int32_t bctbx_ssl_handshake2(bctbx_ssl_context_t *ssl_ctx, void *isRunning);
 BCTBX_PUBLIC void bctbx_ssl_set_io_callbacks(
     bctbx_ssl_context_t *ssl_ctx,
     void *callback_data,
diff --git a/bctoolbox/src/crypto/mbedtls.c b/bctoolbox/src/crypto/mbedtls.c
index 65a7eec..9ed3407 100644
--- a/bctoolbox/src/crypto/mbedtls.c
+++ b/bctoolbox/src/crypto/mbedtls.c
@@ -931,6 +931,11 @@ int32_t bctbx_ssl_read(bctbx_ssl_context_t *ssl_ctx, unsigned char *buf, size_t
 }
 
 int32_t bctbx_ssl_handshake(bctbx_ssl_context_t *ssl_ctx) {
+	return bctbx_ssl_handshake2(ssl_ctx, NULL);
+}
+
+int32_t bctbx_ssl_handshake2(bctbx_ssl_context_t *ssl_ctx, void *isRunning) {
+	mbedtls_ssl_set_user_data_p(&(ssl_ctx->ssl_ctx), isRunning);
 
 	int ret = mbedtls_ssl_handshake(&(ssl_ctx->ssl_ctx));
 
diff --git a/external/mbedtls/library/ssl_tls.c b/external/mbedtls/library/ssl_tls.c
index c773365bf..53e73537b 100644
--- a/external/mbedtls/library/ssl_tls.c
+++ b/external/mbedtls/library/ssl_tls.c
@@ -4607,6 +4607,13 @@ int mbedtls_ssl_handshake(mbedtls_ssl_context *ssl)
 
     /* Main handshake loop */
     while (ssl->state != MBEDTLS_SSL_HANDSHAKE_OVER) {
+        // TN patch - check running flag to abort handshake
+        char *isRunning = mbedtls_ssl_get_user_data_p(ssl);
+        if (isRunning && !*isRunning) {
+            return MBEDTLS_ERR_SSL_CONN_EOF;
+        }
+        // TN patch
+
         ret = mbedtls_ssl_handshake_step(ssl);
 
         if (ret != 0) {
