diff --git a/liblinphone/src/core/platform-helpers/mac-platform-helpers.mm b/liblinphone/src/core/platform-helpers/mac-platform-helpers.mm
index 363afcfed..5154cb84d 100644
--- a/liblinphone/src/core/platform-helpers/mac-platform-helpers.mm
+++ b/liblinphone/src/core/platform-helpers/mac-platform-helpers.mm
@@ -32,6 +32,7 @@
 #include "logger/logger.h"
 #include "core/core.h"
 #include "mac-platform-helpers.h"
+#include "linphone/lpconfig.h"
 
 // =============================================================================
 
@@ -170,9 +171,13 @@
 
 //Get global proxy settings from system and set variables mHttpProxy{Host,Port,Enabled}.
 void MacPlatformHelpers::getHttpProxySettings(void) {
+	// TN patch - use_system_http_proxy is a Linphone config option that allows to use the system HTTP proxy
+	// if set to 1, otherwise it will not use the system HTTP proxy.
+	int useSystemHttpProxy = !!linphone_config_get_int(linphone_core_get_config(getCore()->getCCore()), "sip", "use_system_http_proxy", 0);
+
 	CFDictionaryRef proxySettings = CFNetworkCopySystemProxySettings();
 
-	if (proxySettings) {
+	if (useSystemHttpProxy && proxySettings) {
 		CFNumberRef enabled = (CFNumberRef) CFDictionaryGetValue(proxySettings, kCFNetworkProxiesHTTPEnable);
 		if (enabled != NULL) {
 			int val = 0;
